### PR TITLE
Improve error messages

### PR DESCRIPTION
**Pull request by kluopaja**
~~More complicated API layer error improvements should probably be in a separate task/pull request.~~ Also add more `ApiError`s.

### CODE REVIEW

File: src/service/submit.rs [hunk 1]
```diff
@@ -28,46 +28,55 @@ pub fn update_submit_parameters(
...
-    let task_id = storage
-        .get_task()
-        .ok_or_else(|| anyhow!("Task not provided"))?
-        .to_owned();
-    let language_name = storage
-        .get_language()
-        .ok_or_else(|| anyhow!("Language not provided"))?
-        .to_owned();
-    let language_option = storage.get_option().map(|t| t.to_owned());
+    (|| -> Result<u64> {
```
Comment (ollpu) [inline src/service/submit.rs:31]: Specifying return type shouldn't be necessary
```suggestion
    (|| {
```

File: src/service/submit.rs [hunk 2]
```diff
@@ -28,46 +28,55 @@ pub fn update_submit_parameters(
...
     submission_id: u64,
     poll: bool,
 ) -> Result<SubmissionInfo> {
-    let storage = res.storage.get();
-    let course_id = storage.get_course().unwrap();
-    let task_id = storage.get_task().unwrap();
-    Ok(res
-        .api
-        .get_submit(require_login(res)?, course_id, task_id, submission_id, poll)?)
+    (|| -> Result<SubmissionInfo> {
```
Comment (ollpu) [inline src/service/submit.rs:73]:
```suggestion
    (|| {
```

File: src/service/submit.rs (hunk 1 above)
Comment (kluopaja) [reply src/service/submit.rs:31]: Leaving the return type out results in compilation error as operator `?` needs to know how to cast the error.

**Review by Valokoodari — APPROVED** [review]

File: src/api/mod.rs [hunk 3]
```diff
@@ -30,8 +30,14 @@ pub enum ApiError {
...
     JsonError(#[from] miniserde::Error),
-    #[error("{}", .0.message)]
-    ResponseError(ErrorResponse),
+    #[error("Invalid credentials")]
+    InvalidCredentialsError,
+    #[error("Invalid API key. Log in again.")]
+    ApiKeyError,
+    #[error("Server error: \"{}\"", .0)]
+    ServerError(String),
+    #[error("Client error: \"{}\"", .0)]
```
Comment (ollpu) [inline src/api/mod.rs:39]: These are the technical names, but do they make sense to the user?

Maybe

```suggestion
    #[error("Internal API error: \"{}\"", .0)]
    ServerError(String),
    #[error("API request failed: \"{}\"", .0)]
```

**Review by H4m5t3r — APPROVED** [review]

**Pull request merged**